File: backend/app/routers/songs.py
from fastapi import APIRouter
from app.schemas.translation import TranslationRequest
from app.schemas.card import CardRequest
from app.services.tracks import search_tracks
from app.services.lyrics import fetch_lyrics
from app.services.translation import translate_lines
from app.services.cards import build_cards

router = APIRouter()

@router.get("/songs/search")
def search(q: str):
    return search_tracks(q)

@router.get("/songs/{track_id}/lyrics")
def lyrics(track_id: int):
    return fetch_lyrics(track_id)

@router.post("/songs/translation")
def translation(req: TranslationRequest):
    return translate_lines(req.lines, req.to_lang)

@router.post("/songs/cards")
def cards(req: CardRequest):
    return build_cards(req)

File: frontend/src/services/songs.ts
import type { Card, CardRequest } from "../types/card";
import type { TrackResult } from "../types/track";
import type { TranslatedLine } from "../types/translation";
import { request } from "./api";

const searchTracks = async (query: string): Promise<TrackResult[]> => {
    return await request(`/songs/search?q=${encodeURIComponent(query)}`)
}

const fetchLyrics = async (trackId: number): Promise<string[]> => {
    return await request(`/songs/${trackId}/lyrics`)
}

const translateLines = async (lines: string[], toLang: string): Promise<TranslatedLine[]> => {
    const options: RequestInit = {
        method: "POST",
        body: JSON.stringify({ lines, toLang })
    }
    return await request(`/songs/translation`, options)
}

const buildCards = async (req: CardRequest): Promise<Card[]> => {
    const options: RequestInit = {
        method: "POST",
        body: JSON.stringify(req)
    }
    return await request("/songs/cards", options)
}

export { searchTracks, fetchLyrics, translateLines, buildCards }